src/smtlibparser.yy

```diff
@@ -352,29 +351,46 @@ bvconst:
 ;
 
 sort:
-   BOOL
-   {
-     $$ = drv.solver()->make_sort(smt::BOOL);
-   }
-   | INT
-   {
-     $$ = drv.solver()->make_sort(smt::INT);
-   }
-   | REAL
-   {
-     $$ = drv.solver()->make_sort(smt::REAL);
-   }
-   | indprefix BITVEC NAT RP
+   SYMBOL
    {
-     $$ = drv.solver()->make_sort(smt::BV, std::stoi($3));
+     smt::Sort res;
+     // check built-in sort kinds first
+     smt::SortKind sk = drv.lookup_sortkind($1);
+     if (sk == smt::NUM_SORT_KINDS)
+     {
+       // got the dedicated null enum
+       // check defined sorts
+       res = drv.lookup_sort($1);
+     }
+     else
+     {
+       res = drv.solver()->make_sort(sk);
+     }
+     $$ = res;
    }
-   | LP ARRAY sort sort RP
+   | indprefix SYMBOL NAT RP
    {
-     $$ = drv.solver()->make_sort(smt::ARRAY, $3, $4);
+     // this one is intended for bit-vectors
+     smt::SortKind sk = drv.lookup_sortkind($2);
+     if (sk == smt::NUM_SORT_KINDS)
+     {
+       // got dedicated null enum
+       yy::parser::error(@2, std::string("Unrecognized sort: ") + $2);
+       YYERROR;
+     }
+     $$ = drv.solver()->make_sort(sk, std::stoi($3));
    }
-   | SYMBOL
+   | LP SYMBOL sort sort RP
```

Comment:
I may be missing something but I think that the current parser does not support sorts that are parametrized by more than two sorts and parametrized uninterpreted sorts in general. For example:

```
(declare-sort U 3)                  
(declare-const x (U Bool Bool Bool))
(check-sat)                         
```

Reply:
That's right. It doesn't support `declare-sort` at all yet actually. I'll add that to the list of features to add. Thanks for pointing that out!